Lib/asyncio/runners.py

```diff
@@ -21,6 +25,10 @@ def run(main, *, debug=False):
     It should be used as a main entry point for asyncio programs, and should
     ideally only be called once.
 
+    Each thread within the threadpool is given a timeout duration of 5
+    minutes. If the thread hasn't finishing joining within that duration,
```

Comment:
@1st1 While writing the documentation, I was thinking that @asvetlov's suggestion to the timeout as an option for ``asyncio.run()`` might be beneficial. Although it would add an additional knob for users to adjust, forcing them to strictly use a 5 minute timeout duration seems a bit restrictive.

Do you think it might be useful to give users the option to adjust the timeout duration by adding a *timeout* parameter to `asynio.run()`? Since you regularly interact with daily users of asyncio, you might have a better idea of whether or not that would be useful.

Reply:
I think that adding knobs is only justified when there's request for that from users. In this particular case it's a bit tricky.

The way I think about it is that it's easy to implement whatever timeout is necessary with the following pattern:

```python
asyncio.run(main())

# versus

async def runner():
  try:
    return main()
  finally:
    loop = asyncio.get_running_loop()
    await loop.shutdown_default_executor(3600)

asyncio.run(runner())
```

The argument for this pattern is that in more or less advanced asyncio applications there's always a "runner()" like in the above.  It usually sets up signal handlers, manages global state, etc.  `asyncio.run()` in those cases is simply providing a set of reasonable defaults on how you run your event loop and how you shut it down.  And 5 minutes is a reasonable default in this case, I guess.

Long story short, I'd go with this design and learn from our users if we need to make it configurable.

This is just my opinion on this and I'd love to hear what @asvetlov thinks about it.

Reply:
> The argument for this pattern is that in more or less advanced asyncio applications there's always a "runner()" like in the above.

Ah, I actually hadn't thought as much about how relatively simple it would be for the users to create their own runner. I wonder if it would be worthwhile to document this pattern somewhere in `asyncio-task.rst`, for those who desire further customization.

>  And 5 minutes is a reasonable default in this case, I guess.

Yeah 5 minutes seems reasonable for most tasks. If it's an issue, we can easily adjust it by modifying the constant.

> Long story short, I'd go with this design and learn from our users if we need to make it configurable.

Good point. It might be a better idea to leave out the parameter initially and then add it in later if the users request it. It would be easy to add a new parameter to the API for `asyncio.run()`, but it's incredibly difficult to remove parameters if feature bloat becomes an issue in the future.

> This is just my opinion on this and I'd love to hear what @asvetlov thinks about it.

Personally I have no issue with either of the options, I think both of them would be viable. After reading your latest response, I'm leaning more in favor of keeping the current iteration though.

The more I've been involved with API design, I've increasingly come to realize that there's frequently some degree of disconnect between what the users actually want and what the developers *think* the users want. It can definitely be worthwhile to start with a minimalist approach and add features based on user feedback, rather than making assumptions. Making too many assumptions can easily lead to feature bloat.